core/src/labels/labels.cpp

```diff
@@ -118,64 +108,185 @@ void Labels::update(const View& _view, float _dt, const std::vector<std::unique_
-                tiles.push_back(_cache->contains(t0->sourceID(), tileID.getChild(2)));
-                tiles.push_back(_cache->contains(t0->sourceID(), tileID.getChild(3)));
+                        // The new label lies within the circle defined by the bbox of l0
+                        if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
+                            l0->skipTransitions();
+                        }
+                    }
+                }
             }
+        }
+    }
+}
 
-            for (const auto& t1 : tiles) {
-                if (!t1) { continue; }
-                for (const auto& style : _styles) {
-                    const auto& m0 = t0->getMesh(*style);
-                    if (!m0) { continue; }
-                    const LabelMesh* mesh0 = dynamic_cast<const LabelMesh*>(m0.get());
-                    if (!mesh0) { continue; }
-                    const auto& m1 = t1->getMesh(*style);
-                    if (!m1) { continue; }
-                    const LabelMesh* mesh1 = static_cast<const LabelMesh*>(m1.get());
-
-                    for (auto& l0 : mesh0->getLabels()) {
-                        if (!l0->canOcclude()) { continue; }
-
-                        for (auto& l1 : mesh1->getLabels()) {
-                            if (!l1 || !l1->canOcclude() || l0->hash() != l1->hash()) {
-                                continue;
-                            }
-                            float d2 = glm::distance2(l0->transform().state.screenPos,
-                                    l1->transform().state.screenPos);
-
-                            // The new label lies within the circle defined by the bbox of l0
-                            if (sqrt(d2) < std::max(l0->dimension().x, l0->dimension().y)) {
-                                l0->skipTransitions();
-                            }
-                        }
+void Labels::checkRepeatGroups(std::vector<TextLabel*>& _visibleSet) const {
+    struct GroupElement {
+        glm::vec2 position;
+        float threshold;
+        size_t hash;
+
+        bool operator==(const GroupElement& _ge) {
+            return _ge.position == position
+                && _ge.threshold == threshold
+                && _ge.hash == hash;
+        };
+    };
+
+    std::map<size_t, std::vector<GroupElement>> repeatGroups;
+
+    auto textLabelIt = _visibleSet.begin();
+    while (textLabelIt != _visibleSet.end()) {
+        auto textLabel = *textLabelIt;
+
+        std::size_t seed = 0;
+
+        hash_combine(seed, textLabel->options().repeatGroup);
+
+        GroupElement element;
+        element.position = textLabel->transform().state.screenPos;
+        element.threshold = textLabel->options().repeatDistance;
+        element.hash = seed;
+
+        auto it = repeatGroups.find(seed);
+        if (it != repeatGroups.end()) {
+            std::vector<GroupElement>& group = repeatGroups[seed];
+
```

Comment:
This should use the found iterator directly, instead of looking up seed again:
std::vector<GroupElement>& group = *it;